db/dbScript.sql

```diff
@@ -0,0 +1,393 @@
+  PRIMARY KEY (id),
+  FOREIGN KEY (created_by) REFERENCES Account(id),
+  UNIQUE (username),
+  UNIQUE (email)
+);
+create index if not exists idx_account_username on Account(username);
+create index if not exists idx_account_email on Account(email);
+create index if not exists idx_account_type on Account("type");
+
+create table if not exists Country
+(
+	id uuid not null default uuid_generate_v4(),
+	name varchar(56) not null,
+	primary key(id),
+	UNIQUE (name)
+);
+create index if not exists idx_country_name on Country("name");
+
+create table if not exists City
+(
+	id uuid not null default uuid_generate_v4(),
+	nelatitude FLOAT NOT null check (nelatitude between -90 and 90), --rectangular approximation of the area of the city
+    nelongitude FLOAT NOT null check (nelongitude between -180 and 180),
+    swlatitude FLOAT NOT null check (swlatitude between -90 and 90),
+    swlongitude FLOAT NOT null check (swlongitude between -180 and 180),
+	name varchar(100) not null,
+	country_id uuid not null,
+	primary key(id),
+	foreign key (country_id) references Country(id)
+);
+create index if not exists idx_city_name_country_id on City("name", country_id);
+
+CREATE TABLE if not exists Gym
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),
+  latitude FLOAT NOT null check (latitude between -90 and 90),
+  longitude FLOAT NOT null check (longitude between -180 and 180),
+  name VARCHAR(80) NOT NULL,
+  external_place_id VARCHAR(50) NOT NULL, --id of the place in Google Places API
+  external_rating NUMERIC(4,2) NOT null check (external_rating between 0 and 5), --rating retrieved from Google
+  external_rating_number INT NOT null check (external_rating_number >= 0),--number of accounts that contributed to the google rating
+  phone_number VARCHAR(15),
+  address VARCHAR(80) NOT NULL,
+  website VARCHAR(255),
+  is_wheelchair_accessible BOOL NOT NULL,
+  membership_price NUMERIC(5,2) check (membership_price >= 0),
+  created_at TIMESTAMP NOT null check (created_at <= NOW()),
+  price_changed_at TIMESTAMP check (price_changed_at >= changed_at and price_changed_at > created_at),
+  changed_at TIMESTAMP check (changed_at > created_at and changed_at <= NOW()),
+  internal_rating NUMERIC(4,2) NOT null check (internal_rating between 0 and 5), --local rating by local users
+  internal_rating_number INT NOT null check (internal_rating_number >= 0), --number of accounts that contributed to the local rating
+  congestion_rating NUMERIC(4,2) NOT null check (congestion_rating between 0 and 5),
+  congestion_rating_number INT NOT null check (congestion_rating_number >= 0),
+  owned_by uuid, --a gym account that manages information about this gym
+  currency_id uuid NOT NULL,
+  city_id uuid not null,
+  PRIMARY KEY (id),
+  FOREIGN KEY (owned_by) REFERENCES Account(id),
+  FOREIGN KEY (currency_id) REFERENCES Currency(id),
```

Comment:
What if a gym accepts two or more currencies? We should consider creating a many-to-many relationship with the **currency** table to support this.

Reply:
Okay, this currency is the official currency of the country. When we request the path between two locations in Google Directions API, we have that currency because it also gives us the travelling price (we do that only for public transport) in that currency.
The countries that have multiple official currencies do not have good Google Maps coverage or are quite unlikely to have gym recommendations requested (for instance, sth like Cambodia). I am not sure that we should create overhead of merging tables just for some potential edge cases that are quite unlikely to be ever implemented or used